File: Top_Interview/StackQueue/09_224_Basic_Calculator/09_224_Basic_Calculator_01.cpp
#include <iostream>
#include <stack>
using namespace std;


/*
Given a string s representing a valid expression, implement a basic calculator to evaluate it, and return the result of the evaluation.

Note: You are not allowed to use any built-in function which evaluates strings as mathematical expressions, such as eval().

 

Example 1:

Input: s = "1 + 1"
Output: 2
Example 2:

Input: s = " 2-1 + 2 "
Output: 3
Example 3:

Input: s = "(1+(4+5+2)-3)+(6+8)"
Output: 23
 

Constraints:

1 <= s.length <= 3 * 105
s consists of digits, '+', '-', '(', ')', and ' '.
s represents a valid expression.
'+' is not used as a unary operation (i.e., "+1" and "+(2 + 3)" is invalid).
'-' could be used as a unary operation (i.e., "-1" and "-(2 + 3)" is valid).
There will be no two consecutive operators in the input.
Every number and running calculation will fit in a signed 32-bit integer.
*/


class Solution {
public:
    int calculate(string s) {
        // Using the Stack to assist the calculate
        stack<int> stk;
        int sign = 1;
        int result = 0;
        int i = 0;

        // Using the while loop to iterate the string
        while(i < s.size()) {
            if(isdigit(s[i])) {
                int num = s[i] - '0';
                while(i + 1 < s.size() && isdigit(s[i + 1])) {
                    num = num * 10 + (s[i + 1] - '0');
                    i++;
                }
                result += num * sign;
            }else if(s[i] == '+') {
                sign = 1;
            }else if(s[i] == '-') {
                sign = -1;
            // Deal with the parenthesis situation
            }else if(s[i] == '(') {
                stk.push(result);
                stk.push(sign);
                result = 0;
                sign = 1;
            }else if(s[i] == ')') {
                // 棧頂是符號
                // 這個動作是，最裡面的括弧，已經完成了相加
                // 將外層的正負號跳出
                // 最後再加上上一層的內容
                result = result * stk.top();
                // 符號移除
                stk.pop();
                // 再拿出數字
                result += stk.top();
                // 數字移除
                stk.pop();
            }
            i++;
        }
        return result;
    }
};

int main() {
    Solution sol;
    string s = "1 + 1";
    int res = sol.calculate(s);
    printf("%d", res);
}

File: Top_Interview/StackQueue/09_224_Basic_Calculator/09_224_Basic_Calculator_01.py
'''
Given a string s representing a valid expression, implement a basic calculator to evaluate it, and return the result of the evaluation.

Note: You are not allowed to use any built-in function which evaluates strings as mathematical expressions, such as eval().

Example 1:

Input: s = "1 + 1"
Output: 2
Example 2:

Input: s = " 2-1 + 2 "
Output: 3
Example 3:

Input: s = "(1+(4+5+2)-3)+(6+8)"
Output: 23
 

Constraints:

1 <= s.length <= 3 * 105
s consists of digits, '+', '-', '(', ')', and ' '.
s represents a valid expression.
'+' is not used as a unary operation (i.e., "+1" and "+(2 + 3)" is invalid).
'-' could be used as a unary operation (i.e., "-1" and "-(2 + 3)" is valid).
There will be no two consecutive operators in the input.
Every number and running calculation will fit in a signed 32-bit integer.

'''


class Solution:
    def calculate(self, s: str) -> int:
        # Using a Stack to assist the calculator
        stack = []
        # Using a Sign to note the operation
        # Define sign 1 == '+'
        # Define sign -1 == '-' 
        sign = 1
        result = 0
        i = 0

        # Using for-loop to travel the string
        while i < len(s):
            character = s[i]
            if character.isdigit():
                num = int(s[i])
                while i + 1 < len(s) and s[i + 1].isdigit():
                    num = num * 10 + int(s[i + 1])
                    i += 1
                result += num * sign
            elif character == '+':
                sign = 1
            elif character == '-':
                sign = -1
            elif character == '(':
                stack.append(result)
                stack.append(sign)
                result = 0
                sign = 1
            elif character == ')':
                result = result * stack.pop() + stack.pop()
            i += 1
            
        return result

test = "123"
res = Solution().calculate(test)
print(res)
